crates/payload/builder/src/optimism.rs

```diff
@@ -152,6 +162,126 @@ impl PayloadBuilderAttributes for OptimismPayloadBuilderAttributes {
+#[derive(Debug, Clone)]
+pub struct OptimismBuiltPayload {
+    /// Identifier of the payload
+    pub(crate) id: PayloadId,
+    /// The built block
+    pub(crate) block: SealedBlock,
+    /// The fees of the block
+    pub(crate) fees: U256,
+    /// The blobs, proofs, and commitments in the block. If the block is pre-cancun, this will be
+    /// empty.
+    pub(crate) sidecars: Vec<BlobTransactionSidecar>,
+    /// The rollup's chainspec.
+    pub(crate) chain_spec: Arc<ChainSpec>,
+    /// The payload attributes.
+    pub(crate) attributes: OptimismPayloadBuilderAttributes,
+}
+
+// === impl BuiltPayload ===
+
+impl OptimismBuiltPayload {
+    /// Initializes the payload with the given initial block.
+    pub fn new(
+        id: PayloadId,
+        block: SealedBlock,
+        fees: U256,
+        chain_spec: Arc<ChainSpec>,
+        attributes: OptimismPayloadBuilderAttributes,
+    ) -> Self {
+        Self { id, block, fees, sidecars: Vec::new(), chain_spec, attributes }
+    }
+
+    /// Returns the identifier of the payload.
+    pub fn id(&self) -> PayloadId {
+        self.id
+    }
+
+    /// Returns the built block(sealed)
+    pub fn block(&self) -> &SealedBlock {
+        &self.block
+    }
+
+    /// Fees of the block
+    pub fn fees(&self) -> U256 {
+        self.fees
+    }
+
+    /// Adds sidecars to the payload.
+    pub fn extend_sidecars(&mut self, sidecars: Vec<BlobTransactionSidecar>) {
+        self.sidecars.extend(sidecars)
+    }
+}
+
+impl BuiltPayload for OptimismBuiltPayload {
+    fn block(&self) -> &SealedBlock {
+        &self.block
+    }
+
+    fn fees(&self) -> U256 {
+        self.fees
+    }
+}
+
+impl<'a> BuiltPayload for &'a OptimismBuiltPayload {
+    fn block(&self) -> &SealedBlock {
+        (**self).block()
+    }
+
+    fn fees(&self) -> U256 {
+        (**self).fees()
+    }
+}
+
+// V1 engine_getPayloadV1 response
+impl From<OptimismBuiltPayload> for ExecutionPayloadV1 {
+    fn from(value: OptimismBuiltPayload) -> Self {
+        try_block_to_payload_v1(value.block)
+    }
+}
+
+// V2 engine_getPayloadV2 response
+impl From<OptimismBuiltPayload> for ExecutionPayloadEnvelopeV2 {
+    fn from(value: OptimismBuiltPayload) -> Self {
+        let OptimismBuiltPayload { block, fees, .. } = value;
+
+        ExecutionPayloadEnvelopeV2 {
+            block_value: fees,
+            execution_payload: convert_block_to_payload_field_v2(block),
+        }
+    }
+}
+
+impl From<OptimismBuiltPayload> for OptimismExecutionPayloadEnvelopeV3 {
+    fn from(value: OptimismBuiltPayload) -> Self {
+        let OptimismBuiltPayload { block, fees, sidecars, chain_spec, attributes, .. } = value;
+
+        let parent_beacon_block_root =
+            if chain_spec.is_cancun_active_at_timestamp(attributes.timestamp()) {
+                attributes.parent_beacon_block_root().unwrap_or_else(FixedBytes::<32>::default)
+            } else {
+                panic!("V3 is only possible after cancun");
```

Comment:
I think we also need to set the zero hash here (pre cancun):

> Starting at Ecotone, the parentBeaconBlockRoot must be set to the L1 origin parentBeaconBlockRoot, or a zero bytes32 if the Dencun functionality with parentBeaconBlockRoot is not active on L1.

https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/exec-engine.md#engine_getpayloadv3

but perhaps we should treat this as an error and use TryFrom impl

Reply:
Ok, for now I have set the zero hash for the pre cancun case. If we want to impl TryFrom, what should be the error type?